trin-core/src/portalnet/overlay.rs

```diff
@@ -152,20 +145,229 @@ impl Overlay {
+        &self,
+        content_key: Vec<u8>,
+        enr: Enr,
+    ) -> Result<Vec<u8>, String> {
+        let msg = FindContent { content_key };
+        self.discovery
+            .write()
+            .await
+            .send_talkreq(
+                enr,
+                PROTOCOL.to_string(),
+                Message::Request(Request::FindContent(msg)).to_bytes(),
+            )
+            .await
+    }
+}
+
+/// History network layer on top of the overlay protocol. Encapsulates history network specific data and logic.
+#[derive(Clone)]
+pub struct HistoryNetwork {
+    pub overlay: Arc<OverlayProtocol>,
+}
+
+/// State network layer on top of the overlay protocol. Encapsulates state network specific data and logic.
+#[derive(Clone)]
+pub struct StateNetwork {
+    pub overlay: Arc<OverlayProtocol>,
+}
+
+impl StateNetwork {
+    pub async fn new(
+        discovery: Arc<RwLock<Discovery>>,
+        db: Arc<DB>,
+        portal_config: PortalnetConfig,
+    ) -> Result<(Self, PortalnetEvents), String> {
+        let config = OverlayConfig::default();
+        let kbuckets = Arc::new(RwLock::new(KBucketsTable::new(
+            discovery.read().await.local_enr().node_id().into(),
+            config.bucket_pending_timeout,
+            config.max_incoming_per_bucket,
+            config.table_filter,
+            config.bucket_filter,
+        )));
+        let data_radius = Arc::new(RwLock::new(portal_config.data_radius));
+
+        let protocol_receiver = discovery
+            .write()
+            .await
+            .discv5
+            .event_stream()
+            .await
+            .map_err(|e| e.to_string())
+            .unwrap();
+
+        let overlay = OverlayProtocol {
+            discovery: Arc::clone(&discovery),
+            data_radius,
+            kbuckets,
+        };
+
+        let overlay = Arc::new(overlay);
+
+        let utp_listener = UtpListener {
+            discovery: Arc::clone(&discovery),
+            utp_connections: HashMap::new(),
+        };
+
+        let events = PortalnetEvents {
+            discovery: Arc::clone(&discovery),
+            overlay: Arc::clone(&overlay),
+            protocol_receiver,
+            db,
+            utp_listener,
+        };
+
+        let proto = Self {
+            overlay: Arc::clone(&overlay),
+        };
+
+        Ok((proto, events))
+    }
+
+    /// Convenience call for testing, quick way to ping bootnodes
+    pub async fn ping_bootnodes(&mut self) -> Result<(), String> {
+        // Trigger bonding with bootnodes, at both the base layer and portal overlay.
+        // The overlay ping via talkreq will trigger a session at the base layer, then
+        // a session on the (overlay) portal network.
+        for enr in self
+            .overlay
+            .discovery
+            .read()
+            .await
+            .discv5
+            .table_entries_enr()
+        {
+            debug!("Pinging {} on portal network", enr);
```

Comment:
I guess these logs can be more specific now, like:
```suggestion
            debug!("Pinging {} on portal state network", enr);
```